byterun/caml/memory.h

```diff
@@ -80,6 +80,37 @@ color_t caml_allocation_color (void *hp);
   DEBUG_clear ((result), (wosize));                                         \
 }while(0)
 
+
+#define CAML_REPLACE_FORWARD_NO_LONG 1
+#define CAML_REPLACE_FORWARD_IS_FORWARD 2
+/* Short-cut the value if it points to a forward tag and it is possible
+   By default return true only if the shortcut occured.
+
+   If the following flag is set:
+     - CAML_REPLACE_FORWARD_NO_LONG: shortcut only
+         when the destination pointer is not a long
+     - CAML_REPLACE_FORWARD_IS_FORWARD: return true
+         whenever the value is a forward tag
+ */
+static inline int caml_replace_forward(value *v, int flags){
+  if (Tag_val (*v) != Forward_tag){
+    return 0;
+  } else {
+    value fv = Forward_val (*v);
+    if (((flags & CAML_REPLACE_FORWARD_NO_LONG) && Is_long(fv)) ||
+        (Is_block (fv)
+         && (!Is_in_value_area(fv) || Tag_val (fv) == Forward_tag
+             || Tag_val (fv) == Lazy_tag || Tag_val (fv) == Double_tag))){
+      /* Do not short-circuit the pointer. */
+      return flags & CAML_REPLACE_FORWARD_IS_FORWARD;
+    } else {
+      *v = fv;
```

Comment:
You need to add `v` to the ref_table if it was pointing to the major heap and now points to the minor heap. AFAIR, this is a long-standing bug that was fixed in the gc_patches branch, so let's be careful not to reintroduce it. See lines 226-227 in commit 0225ca01e39289ce1801fb09cd011cdbfb542b8d.